kvdb-rocksdb/src/lib.rs

```diff
@@ -561,40 +536,32 @@ impl Database {
 					})
 					.collect::<Vec<_>>();
 				overlay_data.sort();
+				overlay_data
+			};
 
-				let iter = col.map_or_else(
-					|| db.iterator_opt(IteratorMode::Start, &self.read_opts),
-					|c| {
-						db.iterator_cf_opt(cfs[c as usize], IteratorMode::Start, &self.read_opts)
-							.expect("iterator params are valid; qed")
-					},
-				);
-
-				Some(DatabaseIterator { iter: interleave_ordered(overlay_data, iter), _marker: PhantomData })
-			}
-			None => None,
-		}
+			let guarded = iter::ReadGuardedIterator::new(read_lock, col);
+			Some(interleave_ordered(overlay_data, guarded))
+		} else {
+			None
+		};
+		optional.into_iter().flat_map(identity)
 	}
-
-	fn iter_from_prefix(&self, col: Option<u32>, prefix: &[u8]) -> Option<DatabaseIterator<'_>> {
-		match *self.db.read() {
-			Some(DBAndColumns { ref db, ref cfs }) => {
-				let iter = col.map_or_else(
-					|| db.iterator_opt(IteratorMode::From(prefix, Direction::Forward), &self.read_opts),
-					|c| {
-						db.iterator_cf_opt(
-							cfs[c as usize],
-							IteratorMode::From(prefix, Direction::Forward),
-							&self.read_opts,
-						)
-						.expect("iterator params are valid; qed")
-					},
-				);
-
-				Some(DatabaseIterator { iter: interleave_ordered(Vec::new(), iter), _marker: PhantomData })
-			}
-			None => None,
-		}
+	/// Get database iterator from prefix for flushed data.
+	/// Will hold a lock until the iterator is dropped
+	/// preventing the database from being closed.
+	fn iter_from_prefix<'a>(
+		&'a self,
+		col: Option<u32>,
+		prefix: &[u8],
+	) -> impl Iterator<Item = iter::KeyValuePair> + 'a {
+		let read_lock = self.db.read();
+		let optional = if read_lock.is_some() {
+			let guarded = iter::ReadGuardedIterator::new_from_prefix(read_lock, col, prefix);
+			Some(interleave_ordered(Vec::new(), guarded))
```

Comment:
I wonder whether we should search in overlay column as we do in `iter` (cc @arkpar)?